src/canopy/llm/cohere.py

```diff
@@ -0,0 +1,284 @@
+import time
+from copy import deepcopy
+from typing import Union, Iterable, Optional, Any, Dict, List
+
+import cohere  # type: ignore
+from tenacity import retry, stop_after_attempt
+
+from canopy.llm import BaseLLM
+from canopy.llm.models import Function
+from canopy.models.api_models import (
+    _Choice,
+    _StreamChoice,
+    ChatResponse,
+    StreamingChatChunk,
+    TokenCounts,
+)
+from canopy.models.data_models import Context, MessageBase, Messages, Role, Query
+from canopy.context_engine.context_builder.stuffing import StuffingContextContent
+
+
+class CohereLLM(BaseLLM):
+    """
+    Cohere LLM wrapper built on top of the Cohere Python client.
+
+    Note: Cohere requires a valid API key to use this class.
+          You can set the "CO_API_KEY" environment variable to your API key.
+    """
+    def __init__(self,
+                 model_name: str = "command-nightly",
+                 *,
+                 api_key: Optional[str] = None,
+                 base_url: Optional[str] = None,
+                 **kwargs: Any,
+                 ):
+        """
+        Initialize the Cohere LLM.
+
+        Args:
+            model_name: The name of the model to use. See https://docs.cohere.com/docs/models
+            api_key: Your Cohere API key. Defaults to None (uses the "CO_API_KEY" environment variable).
+            base_url: The base URL to use for the Cohere API. Defaults to None (uses the "CO_API_URL" environment variable if set, otherwise use default Cohere API URL).
+            **kwargs: Generation default parameters to use for each request. See https://platform.openai.com/docs/api-reference/chat/create
+                    For example, you can set the temperature, p, etc
+                    These params can be overridden by passing a `model_params` argument to the `chat_completion` methods.
+        """  # noqa: E501
+        super().__init__(model_name)
+        self._client = cohere.Client(api_key, api_url=base_url)
```

Comment:
@igiloh-pinecone I have added the `try`-`except` here as requested. I did the test a little bit different from the equivalent for OpenAI - using the pytest `monkeypatch` fixture to set the env var instead of a custom `bad_api_key` fixture.  I think this way is more elegant but I can change it to do it the other way for consistency if you'd like.